### PR TITLE
fix(provider): Anthropic providers send x-api-key not Bearer for /v1/models

### PR DESCRIPTION
**Pull request by chocobar**
## Problem

Adding Anthropic as a provider via the UI fails with:

```
upstream models endpoint unreachable: failed to get models from
'https://api.anthropic.com/v1/models' provider: 401 Unauthorized -
{"type":"error","error":{"type":"authentication_error","message":"Invalid bearer token"}}
```

Zed agents using that provider also fail to authenticate. "Invalid bearer token" is the tell: Anthropic only emits that when an `Authorization: Bearer` header reaches it. API keys must go in `x-api-key`.

## Root cause

`RetryableClient.ListModels` dispatches to `listAnthropicModels` (which sends `x-api-key`) only when `c.isAnthropic` is true; otherwise it falls through to `listOpenAIModels` (which sends `Authorization: Bearer`).

The env/config provider path sets the flag (`SetIsAnthropic(true)`), but `initializeClient` — the DB/UI-configured provider path — never did. This regressed in commit `c6f57128f` which removed the `isAnthropicProvider(c.baseURL)` base-URL fallback from the dispatch, leaving the DB path with no way to be detected as Anthropic.

## Fix

Set `isAnthropic` in `initializeClient` for endpoints that target Anthropic (name `anthropic`, matching the UI's canonical name, or an `api.anthropic.com` base URL). Vertex endpoints are excluded since they authenticate via OAuth, not `x-api-key`.

## Test

`anthropic_models_regression_test.go` captures the auth header via httptest: without the flag the client sends `Authorization: Bearer` (the bug); with it, `x-api-key` (the fix).

NOT yet tested end-to-end against a live stack — needs `./stack rebuild api` then re-adding the provider in the UI.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by chocobar** [comment]
## PR Review: fix(provider): Anthropic providers send x-api-key not Bearer for /v1/models

### Verdict: Approve with minor suggestions

The fix is correct and well-targeted. Root cause is clear, two fix sites are covered, and test coverage is good.

### What's good
- Root cause traced to `c6f57128f` removing URL-sniffing from the dispatch — clean diagnosis
- Both `initializeClient` (DB/UI path) and `updateClientAPIKeyFromFile` (key rotation path) get `SetIsAnthropic(true)`. Good catch on the rotation path — without it, a key refresh would silently revert to Bearer auth
- Vertex exclusion is correct (OAuth, not x-api-key)
- Tests are well-structured: unit tests for `isAnthropicAPIEndpoint` edge cases + integration test through `initializeClient`
- The ponytail comment honestly flags the heuristic nature of the detection

### Concerns / suggestions

**1. The "renamed proxy" gap is worth tracking with a follow-up issue.** The test case `{"renamed proxy (known gap)", ...}` documents that `isAnthropicAPIEndpoint` won't match a custom-named endpoint pointing at a non-`api.anthropic.com` URL (e.g., an internal proxy). Anyone using an Anthropic proxy with a custom name/base URL will still hit the 401 — they'd need to name their endpoint `anthropic` or include `api.anthropic.com` in the URL. Not a blocker, but worth a follow-up issue for the proper `api_format`/`provider_kind` field on `ProviderEndpoint`.

**2. `strings.Contains(strings.ToLower(endpoint.BaseURL), "api.anthropic.com")` could match false positives.** E.g. `https://api.anthropic.com.proxy.mycompany.com/v1` would match. A more precise check would verify the URL's host is exactly `api.anthropic.com`. Unlikely edge case and the ponytail comment already flags it — up to you whether to tighten now.

**3. Does the chat completion path also need `isAnthropic` for `x-api-key`?** The fix is scoped to `/v1/models` (the `ListModels` dispatch). Does the go-openai client also send `Authorization: Bearer` for chat completions against Anthropic? If Anthropic accepts Bearer for `/v1/messages` but not `/v1/models`, the current fix is sufficient. Worth confirming this isn't half-fixing the problem.

**4. E2E test.** The PR body says "NOT yet tested end-to-end against a live stack" — per project rules, this should be tested. An E2E smoke test (add Anthropic provider in the UI → verify models load) would close the loop.

File: api/pkg/openai/manager/provider_manager.go [hunk 1]
```diff
@@ -420,6 +425,23 @@ func (m *MultiClientManager) GetClient(_ context.Context, req *GetClientRequest)
...
+// isAnthropicAPIEndpoint reports whether a DB/UI-configured provider endpoint
+// targets the direct Anthropic API and therefore needs x-api-key auth (not
+// Bearer) for /v1/models. Detection keys off the canonical provider name
+// (matching the frontend's `name === 'anthropic'` convention) or an
+// api.anthropic.com base URL. Vertex endpoints are excluded — they authenticate
+// via OAuth, not x-api-key.
+// ponytail: name/URL heuristic; the proper fix is an explicit provider-kind /
+// api-format field on ProviderEndpoint, which would also catch Anthropic
+// endpoints renamed and pointed at a proxy (not matched here today).
+func isAnthropicAPIEndpoint(endpoint *types.ProviderEndpoint) bool {
```
Comment (chocobar) [inline api/pkg/openai/manager/provider_manager.go:438]: Nit: `strings.Contains(strings.ToLower(endpoint.BaseURL), "api.anthropic.com")` could match false positives like `https://api.anthropic.com.proxy.mycompany.com/v1`. Consider parsing the URL and checking `host == "api.anthropic.com"` instead. Low priority since the ponytail comment already flags the heuristic, and this is an unlikely edge case.

File: api/pkg/openai/manager/provider_manager.go [hunk 2]
```diff
@@ -286,6 +286,11 @@ func (m *MultiClientManager) updateClientAPIKeyFromFile(provider types.Provider,
 		TLSSkipVerify: m.cfg.Tools.TLSSkipVerify,
 	})
 
+	// Preserve x-api-key auth across key rotation for the Anthropic global provider.
+	if provider == types.ProviderAnthropic {
+		openaiClient.SetIsAnthropic(true)
+	}
```
Comment (chocobar) [inline api/pkg/openai/manager/provider_manager.go:293]: Good catch adding this here too. Without it, a key refresh via `updateClientAPIKeyFromFile` would recreate the client without `isAnthropic`, silently reverting to Bearer auth.

One thing to note: this guard checks `provider == types.ProviderAnthropic`, which only covers the global (env/config) provider path. DB/UI endpoints don't go through `updateClientAPIKeyFromFile` (they go through `initializeClient`), so this is correct as-is. Just calling it out for future reference in case the rotation path ever gets used for DB endpoints.

File: api/pkg/openai/anthropic_models_regression_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+		gotAuth = r.Header.Get("Authorization")
+		gotAPIKey = r.Header.Get("x-api-key")
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write([]byte(`{"data":[{"id":"claude-opus-4-1","display_name":"Claude","type":"model"}]}`))
+	}))
+	defer ts.Close()
+
+	t.Run("without SetIsAnthropic sends Bearer (the bug)", func(t *testing.T) {
+		client := NewWithOptions("sk-ant-test", ts.URL+"/v1", false, ClientOptions{})
+		_, _ = client.ListModels(context.Background())
```
Comment (chocobar) [inline api/pkg/openai/anthropic_models_regression_test.go:34]: Minor footgun: `gotAuth` and `gotAPIKey` are shared closure variables reset between subtests. This works because subtests are sequential, but if anyone adds `t.Parallel()` or a third subtest, it breaks. Consider scoping the captured values per-subtest (e.g., each subtest creates its own `httptest.NewServer`). Low priority.

**Comment by chocobar** [comment]
Thanks @chocobar. Addressed in 58e96390c:

**2. Base-URL false positives** - Fixed. Switched from substring to host-exact: `url.Parse(BaseURL).Hostname() == "api.anthropic.com"` (case-insensitive). `https://api.anthropic.com.proxy.evil.com/v1` now correctly does NOT match; added a test case for it.

**3. Does the chat path also need this?** - Confirmed it does not; the fix isn't half-done. Anthropic has no `/v1/chat/completions`, so `/v1/models` is the only place the go-openai client talks to Anthropic directly. Chat/inference goes through `/v1/messages` -> `anthropicAPIProxyHandler` -> `anthropic.Proxy`, which already sets `x-api-key` from `endpoint.APIKey` (`api/pkg/anthropic/anthropic_proxy.go:219-221`). So inference auth was always correct; only the model-list fetch was on the Bearer path.

**1. Renamed-proxy gap** - Agreed it's worth a proper fix (explicit `provider_kind`/`api_format` field on `ProviderEndpoint`). Tracked in our design/TODO backlog rather than a GitHub issue (team convention). The ponytail comment and the `renamed proxy (known gap)` test case document it in-tree.

**Regotcha (inline on the regression test)** - Fixed: captured headers are now scoped per-subtest via a closure, so adding `t.Parallel()` or a third subtest can't leak captures.

**4. E2E** - Fair. I couldn't run a live UI e2e from my environment (no local docker stack available). What is covered: `Test_InitializeClient_AnthropicAuth` runs a real `ProviderEndpoint` through the actual `initializeClient` production path against an httptest server and asserts `x-api-key` vs Bearer per case (incl. Vertex exclusion) - so the wired path is exercised, not just the primitive. A UI smoke test (add Anthropic provider -> models load) still wants doing against a live stack before/at merge.
